test cases/common/229 persubproject options/meson.build

```diff
@@ -1,10 +1,24 @@
-project('persubproject options', 'c', default_options : ['default_library=both'])
+project('persubproject options', 'c', 'cpp',
+  default_options : [
+    'default_library=both',
+    'c_args=-DPARENT',
+    'cpp_std=c++98'
+  ]
+)
 
 assert(get_option('default_library') == 'both', 'Parent default_library should be "both"')
+assert(get_option('c_args') == ['-DPARENT'], 'Parent c_args are wrong')
 
 # Check it build both by calling a method only both_libraries target implement
 lib = library('lib1', 'foo.c')
 lib.get_static_lib()
 
-subproject('sub1')
-subproject('sub2', default_options : ['default_library=static'])
+sub1 = subproject('sub1')
+sub2 = subproject('sub2',
+  default_options : [
+    'default_library=static',
+    'c_args=-DSUB2_FROM_PARENT',
+  ]
+)
+
+test('mixing-cpp-version', executable('test1', 'main.cpp', link_with : sub2.get_variable('libcpp11')))
```

Comment:
This is UB AFAIK, since it's not safe to mix c++98/03 with c++11/c++14/c++17. It *is* safe to mix  c++11/c++14/c++17 though.
This would've been alright if `sub2` was a dynamic library with C interface.

Reply:
I'm too lazy to look through every compiler, but here is the proof for `gcc`:

> The C++98 language is ABI-compatible with the C++11 language, but several places in the Standard Library break compatibility. This makes it potentially dangerous to link C++98 objects with C++11 objects. 
Source: https://gcc.gnu.org/wiki/Cxx11AbiCompatibility

Reply:
> This would've been alright if `sub2` was a dynamic library with C interface.

That's the intention, yes. I think it's fine because void foo(void) is simple enough to have the same ABI, but it should have `extern "C"` to make it explicit.

Reply:
In this test it's built and linked as a static library `default_library=static`, and there is `std::shared_ptr` in `sub2/foo.cpp`. 

Either way, it seems that my gcc knowledge on this subject is flawed, since there is a direct confirmation of c++03/c++11+ ABI compatibility by `libstdc++` author: https://stackoverflow.com/a/49119902/6310318

Reply:
> In this test it's built and linked as a static library `default_library=static`, and there is `std::shared_ptr` in `sub2/foo.cpp`.

Oh good point, haven't noticed that was static linked. Needs to be fixed indeed.

> Either way, it seems that my gcc knowledge on this subject is flawed, since there is a direct confirmation of c++03/c++11+ ABI compatibility by `libstdc++` developer: https://stackoverflow.com/a/49119902/6310318

That's gcc specific, if I understand correctly. Wouldn't be surprised that's in the category of undefined in the spec but all compilers actually keep compat.

Reply:
> Wouldn't be surprised that's in the category of undefined in the spec but all compilers actually keep compat.

C++ standard doesn't know anything about ABI at all :D
Regarding other compilers: clang claims to be ABI compatible with GCC (thus theoretically it should handle C++XX compatibility the same); MSVC is ABI compatible across C++14/C++17 (https://www.reddit.com/r/cpp/comments/alsoqx/c_binary_compatibility_and_painfree_upgrades_to/efhj7kp)